### PR TITLE
Add Message Input/Output & receipt

### PR DESCRIPTION
**Pull request by vlopes11**
Include message input and output with their respective validation rules.

This commit will introduce a stateful property to the transaction
validation. A valid transaction might not be validated anymore in
post-execution since its outputs might be mutated at runtime in the VM.

Related issue: https://github.com/FuelLabs/fuel-specs/pull/318

### CODE REVIEW

**Comment by vlopes11** [comment]
Blocked by https://github.com/FuelLabs/fuel-asm/pull/78

File: src/transaction/validation.rs
Comment (adlerjohn) [inline src/transaction/validation.rs:157]: > producing invalid transactions.

If the VM is implemented correctly (we can assume it is), then a valid transaction that's input into the VM will output a valid transaction. Anything else would be a DoS vector.

File: src/transaction/validation.rs
Comment (adlerjohn) [inline src/transaction/validation.rs:171]: Why are zero-valued address and amount forbidden for message outputs? They're like variable outputs, they can be zero and if they're zero then they're not included in the UTXO set.

File: src/transaction/validation.rs
Comment (vlopes11) [reply src/transaction/validation.rs:157]: We do override the recipient and amount of the output:

https://github.com/FuelLabs/fuel-specs/blob/e0d5d8aea3f6c344ca47776403accc58b67eeb01/specs/vm/instruction_set.md?plain=1#L1417-L1420

It sounds odd that we execute a transaction (with zeroed recipient/amount) different than the one that was signed (with original recipient/amount). If a transaction is submitted with these fields filled, shouldn't it be invalid?

File: src/transaction/validation.rs
Comment (vlopes11) [reply src/transaction/validation.rs:171]: We can't execute `SMO` if the recipient is non-zero. Isn't this an invalid tx? One that cannot ever have its output consumed by its respective opcode?

https://github.com/FuelLabs/fuel-specs/blob/e0d5d8aea3f6c344ca47776403accc58b67eeb01/specs/vm/instruction_set.md?plain=1#L1401

Comment (pixelcircuits) [reply src/transaction/validation.rs:171]: I agree this validation check is necessary, but I'm confused why there isn't a similar check for OutputType.Variable to verify that amount is zero.

File: src/transaction/validation.rs
Comment (adlerjohn) [reply src/transaction/validation.rs:157]: It should be invalid, yes, but why is this stateful is my confusion.

Comment (vlopes11) [reply src/transaction/validation.rs:157]: This is ofc a matter of perspective, but that would be the "pre-execution" and "post-execution" states of a transaction. I will alter the behavior to produce valid transactions for both conditions and just force these values to zero during VM initialization (as it is literally described in the specs).

I just feel a bit uncomfortable of mutating the tx during VM initialization since it creates a magic box effect

Comment (vlopes11) [reply src/transaction/validation.rs:157]: https://github.com/FuelLabs/fuel-tx/pull/149/commits/411a630c2ecacba80711c038d65e03afa1506262

File: src/receipt.rs
Comment (Voxelot) [inline src/receipt.rs:165]: Should we be publicly exposing this function that doesn't guarantee the length is padded correctly?

File: src/transaction.rs
Comment (Voxelot) [inline src/transaction.rs:464]: Should this also zero out variable outputs?

File: src/transaction/consensus_parameters.rs
Comment (Voxelot) [inline src/transaction/consensus_parameters.rs:481]: Since these are deprecated, we probably don't want to add new constants here?

File: src/transaction/id.rs
Comment (Voxelot) [inline src/transaction/id.rs:52]: is this duplicating `prepare_init`?

File: src/transaction/internals.rs
Comment (Voxelot) [inline src/transaction/internals.rs:53]: any reason why these should be gated as internal? Seems useful if we wanted to have the ability to edit these fields without big match blocks around the tx in the VM or client for malleation.

File: src/transaction/validation.rs
Comment (Voxelot) [inline src/transaction/validation.rs:257]: Do we need to check for duplicate message inputs separately here since messages don't have a UTXO ID?

File: src/transaction/consensus_parameters.rs
Comment (adlerjohn) [reply src/transaction/consensus_parameters.rs:481]: Why is message data deprecated?

File: tests/prepared_init.rs
Comment (Voxelot) [inline tests/prepared_init.rs:6]: Since other outputs or parts of the tx may be zeroed in the future, should we clarify this test is only for message outputs?

```suggestion
fn output_prepare_init_zeroes_message_recipient_and_amount() {
```

File: tests/valid_cases/input.rs
Comment (Voxelot) [inline tests/valid_cases/input.rs:261]: Needs docs clarifying what the expected failure actually is here

```suggestion
        .expect("Expected failure");
```

**Review by Voxelot — CHANGES_REQUESTED** [review]
Only blocking feedback from me is clarification on whether we are properly testing for duplicate message inputs (i.e. double spend).

File: src/receipt.rs
Comment (vlopes11) [reply src/receipt.rs:165]: This one will allow `const` directives and is used to restore raw serialization. It means the output will always be the representation of what is serialized, without making assumptions.

File: src/transaction.rs
Comment (vlopes11) [reply src/transaction.rs:464]: https://github.com/FuelLabs/fuel-tx/pull/149/commits/38ebb1a7d60015f6d2c6cda669eb71eb2a53783d

File: src/transaction/consensus_parameters.rs
Comment (vlopes11) [reply src/transaction/consensus_parameters.rs:481]: Its important we always keep consistency. If we are to remove these constants, we should remove all of them in one release. The user will still receive the deprecated warning tho.

@adlerjohn we are using the `ConsensusParameter` struct to fetch these values. These constants will be removed soon.

File: src/transaction/id.rs
Comment (vlopes11) [reply src/transaction/id.rs:52]: Its a different context. Even tho we do the same, here its related to the signature, not to the VM initialization.

File: src/transaction/internals.rs
Comment (vlopes11) [reply src/transaction/internals.rs:53]: We should use the methods and API instead, that will (or at least should) guarantee tx consistency.

Exposing the internals is not only allowing the users to use the API to create invalid transactions, but will make their code more vulnerable to breaking changes. Ideally, somewhere in the future, we should never have to `match` these enums and use only the available API.

However, if for some reason the user still wants to manipulate the shielded attributes of a transaction, he can just, knowingly, activate this feature

File: src/transaction/validation.rs
Comment (vlopes11) [reply src/transaction/validation.rs:257]: The output of `Input::utxo_id` is now `Option<&UtxoId>` instead of `&UtxoId`. This means the check will be executed only for inputs that contains utxo id

File: src/transaction/id.rs
Comment (Voxelot) [reply src/transaction/id.rs:52]: Makes sense, clearing for signatures isn't the same as for init (eg witnesses etc)

**Review by Voxelot — CHANGES_REQUESTED** [review]
To be specific, I think we should have a `ValidationError::TransactionInputMessageIdDuplicated` error and prevent the same message ID from being included twice on the same transaction. The other inputs types are covered by checks on utxoid & contract id.

edit: I think we don't need checks against duplicate output message ids, since those are derived instead of stored / serialized.

**Comment by adlerjohn** [comment]
Preventing the same message from being spent twice on the input side is required, otherwise it can be double-spent. Output messages as @Voxelot said derive their ID from the tx ID + output index, so by construction are unique.

**Comment by vlopes11** [comment]
It doesn't look like this validation is trivial. Taking the asset id as example, we don't cross-check duplicated asset ids, as in:

https://github.com/FuelLabs/fuel-vm/blob/94810b05f3a2644e9b9a05a6bef604de84bb1250/src/interpreter/initialization.rs#L93-L101

We can generate different message id if we use the same sender but different recipient - so this suggests we can't reduce the validation to duplicated message id. I suggest we open a dedicated issue to discuss that, update the specs and after update the code here.

**Comment by Voxelot** [comment]
> we don't cross-check duplicated asset ids

We are trying to mitigate the risk of users including the same UTXO / spendable input multiple times, thus effectively permitting double-spending. Duplicated asset ids across coin inputs are fine, as long as they are distinct coins. I'm having difficulty understanding the problem in the init code you referenced, as this is a reduce-operation into a hashmap (guaranteed uniqueness) which accumulates value (+=) across matching asset ids.

My understanding is that the message id is already committed to by L1/bridging, and isn't malleable. Afaik the exact specifics of our bridging haven't been made public yet. However, there is internal documentation already available from the bridging team. 

To shorten the cycle here, I'm just going to push the necessary changes to quickly unblock this.

My followup question however, is if we should also have validation that the message id claimed on the input actually matches the other fields. Since the client will likely just validate the message id exists in the state set, but won't check its validity.

**Comment by adlerjohn** [comment]
The bridge contract on the other chain is guaranteed to construct unique message IDs for each message, and the client is guaranteed to only insert valid messages into its message set. Checking that the message ID matches the fields is redundant due to that.

**Review by adlerjohn — APPROVED** [review]

**Review by Voxelot — APPROVED** [review]

**Comment by pixelcircuits** [comment]
Just FYI, the Ethereum contract doesn't calculate the input messageId. As a gas saving measure, it emits the data required to compute the messageId, but leaves the Fuel client to calculate the actual messageId.

**Pull request merged**
